### DISCUSSION TITLE
Electronic Invoicing (CFDI) Module for Mexico

### DISCUSSION THREAD

**Discussion by azubieta** [OCA/odoo-community.org#85] [General]
I would like to contribute with a module for electronic invoicing for México. I'm aware this is an existent feature on Odoo enterprise but I think that it would be great to have it also for community. So I will be investing some time and effort on this project.

I have some questions about it:
- Are there any guidelines for development OCA modules?
- Is there a defined project structure ?
- Which language should be used by default, my first guess is to use Spanish as this will only be used in Mexico?
- Are there other people working on something similar ?
- I may also need to some advice on how certain features should be implemented. 

Currently I have a project draft at Github were I plan to put my work. Please feel free to share your comments and recommendations here or there:

Github: https://github.com/AugeTec/odoo_cfdi

**Comment by max3903** [comment]
Hello @azubieta 

Happy to help and support in anyway I can.

> Are there any guidelines for development OCA modules?

https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst

> Is there a defined project structure ?

Not really. We can discuss options... One I can think of is:

* l10n_mx_edi_oca
* l10n_mx_edi_account_oca

hosted on https://github.com/OCA/l10n-mexico

Another one was discussed couple years ago here: https://github.com/OCA/l10n-mexico/issues/6

> Which language should be used by default, my first guess is to use Spanish as this will only be used in Mexico?

English please. Translations are managed with Weblate.

> Are there other people working on something similar ?

Not that I know of.

> I may also need to some advice on how certain features should be implemented.

Again happy to help.

**Comment by luisg123v** [comment]
> Is there a defined project structure ?

I noticed your project name containg underscores. The only "guideline" I know of regarding projects is that they should use middle dashes instead of underscores, e.g. `l10n-mx-oca`

**Reply by max3903** [reply]
Module names contains underscores. Repository name contains dashes.

The repository already exists: https://github.com/OCA/l10n-mexico

**Reply by azubieta** [reply]
Regarding [l10n-mexico](https://github.com/OCA/l10n-mexico) should I take it as base or it's fine to start something new ?

**Comment by azubieta** [comment]
Ok, let's use `l10n-mexico` as repository name. I'll try to achieve one goal peer week so you could help me on reviewing it. Notice that I'll be migrating my work on previous module I did for my company Auge Tec. 
The goal for this week will be creating a module for holding the CFDI catalogs.

**Reply by agb80** [reply]
@azubieta we also have a work that would like to share to OCA with a very similar scope than yours as we have discussed early this year with @max3903 

Unfortunately by several extra work that we had during first quarter couldn't start on this until now. Would like to share efforts if you agree, how can we contact to you to coordinate a meeting and create a collaborative work schedule?

**Reply by azubieta** [reply]
@agb80 you can reach me at [contact@azubieta.net](mailto:contact@azubieta.net). All help is welcome :)

**Reply by agb80** [reply]
Send you an email

**Comment by azubieta** [comment]
I just forked OCA/l10n-mexico and would like to create some issues to discuss the implementation details. Should I use the upstream repo or my fork?

Fork at:  https://github.com/AugeTec/l10n-mexico

**Reply by pedrobaeza** [reply]
The issues should be put on the main repo (OCA/l10n-mexico).

**Comment by azubieta** [comment]
For those interested in a more direct chat you can join us at https://discord.gg/rN5xRdE in the room l10n-mexico.

**Comment by azubieta** [comment]
We had our first technical meeting below you will find a summary:
- odoo community 15 will the target
- the scope will be limited to invoice, payments and credit notes
- @azubieta will be working on creating the base models to convert to and form CFDI xml files using the generated schemas.
- @agb80 will be working on implementing the ACP (PAC in Spanish) management/integration base module

- there will be a mixin generated from XSD files the will be implemented by a CFDI module which will be integrated on modules later
![Screenshot_20230414_114227](https://user-images.githubusercontent.com/1138094/232118765-9e2df490-b2c4-4280-b2b3-39bed109ac96.png)

**Comment by azubieta** [comment]
## Project Update

I have been trying to generate the mixins from the CFDI xsd files using xsdata, xsdata-odoo, and spec_driven_models. The idea is to a simple import/export features and base models auto generated. This was achieve partially due to limitations of the tools. To be specific xsdata-odoo currently is a good fit for the Brazilian localization but it don't generated good results with the CFDI specification. Regarding `spec_driven_models` the idea of generating the concrete models from mixins and their access rules on "the fly" makes me "nervous" about future maintenance issues. An unseen bug may become a nightmare due the large amount of "auto-magic" and the leak of tests cases. `xsdata-odoo` also needs some love in order to get issues fixed, such as duplicated model names. While I don't discard the possibility of using these tools in the future I would like to do it with a solid set of test cases.

Take away, we will be using the xsdata generated classes and the generated models will be manually fixed. That will make the l10n_mx_cfdi_spec module. On top of that we will implement the base cfdi document module which integrates the spec with the catalogs.

tldr: `xsdata-odoo` has issues, `spec_driven_model` needs to mature a bit. Taking what could be used from them an moving on.
